cns/deviceplugin/pluginmanager.go

```diff
@@ -0,0 +1,147 @@
+	"context"
+	"fmt"
+	"os"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/Azure/azure-container-networking/crd/multitenancy/api/v1alpha1"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+	"k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1"
+)
+
+var ErrUnknownDeviceType = errors.New("unknown device type")
+
+const (
+	defaultDevicePluginDirectory      = "/var/lib/kubelet/device-plugins"
+	defaultDeviceCheckInterval        = 5 * time.Second
+	defaultServerCountForDevicePlugin = 2 // Number of device types supported
+)
+
+type pluginManagerOptions struct {
+	devicePluginDirectory string
+	kubeletSocket         string
+	deviceCheckInterval   time.Duration
+}
+
+type pluginManagerOption func(*pluginManagerOptions)
+
+func PluginManagerSocketPrefix(prefix string) func(*pluginManagerOptions) {
+	return func(opts *pluginManagerOptions) {
+		opts.devicePluginDirectory = prefix
+	}
+}
+
+func PluginManagerKubeletSocket(socket string) func(*pluginManagerOptions) {
+	return func(opts *pluginManagerOptions) {
+		opts.kubeletSocket = socket
+	}
+}
+
+func PluginDeviceCheckInterval(i time.Duration) func(*pluginManagerOptions) {
+	return func(opts *pluginManagerOptions) {
+		opts.deviceCheckInterval = i
+	}
+}
+
+// PluginManager runs device plugins for vnet nics and ib nics
+type PluginManager struct {
+	Logger        *zap.Logger
+	VnetNICPlugin *Plugin
+	IBNICPlugin   *Plugin
+	options       pluginManagerOptions
+}
+
+func NewPluginManager(l *zap.Logger, initialVnetNICCount, initialIBNICCount int, opts ...pluginManagerOption) *PluginManager {
+	logger := l.With(zap.String("component", "devicePlugin"))
+	socketWatcher := NewSocketWatcher(logger)
+	options := pluginManagerOptions{
+		devicePluginDirectory: defaultDevicePluginDirectory,
+		kubeletSocket:         v1beta1.KubeletSocket,
+		deviceCheckInterval:   defaultDeviceCheckInterval,
+	}
+	for _, o := range opts {
+		o(&options)
+	}
+	return &PluginManager{
+		Logger: logger,
+		VnetNICPlugin: NewPlugin(logger, string(v1alpha1.DeviceTypeVnetNIC), socketWatcher,
+			getSocketName(options.devicePluginDirectory, v1alpha1.DeviceTypeVnetNIC), initialVnetNICCount,
+			options.kubeletSocket, options.deviceCheckInterval),
+		IBNICPlugin: NewPlugin(logger, string(v1alpha1.DeviceTypeInfiniBandNIC), socketWatcher,
+			getSocketName(options.devicePluginDirectory, v1alpha1.DeviceTypeInfiniBandNIC), initialIBNICCount,
+			options.kubeletSocket, options.deviceCheckInterval),
+		options: options,
+	}
+}
+
+// Run runs the plugin manager until the context is cancelled or error encountered
+func (p *PluginManager) Run(ctx context.Context) error {
+	// clean up any leftover state from previous failed plugins
+	// this can happen if the process crashes before it is able to clean up after itself
+	if err := p.cleanOldState(); err != nil {
+		return errors.Wrap(err, "error cleaning state from previous plugin process")
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(defaultServerCountForDevicePlugin) // starting two goroutines to wait on
```

Comment:
I think it's easier to safely handle waitgroups if you always do `wg.Add(1)`. Here, it's no big deal, since you can audit it visually. It's as the code changes over time where it gets more dangerous.

Reply:
It's the linter complaining. :(

Reply:
updated

Reply:
Sorry, I meant literally `wg.Add(1)` (i.e. the only acceptable argument to `wg.Add` is 1). That way you can be more sure that the goroutines that use the waitgroup are paired up with appropriate increments and decrements.

Reply:
Done